File: daily/Backtracking/17-Letter-combinations-of-a-phone-number.go
package daily

func letterCombinations(digits string) []string {
	res := []string{}
	if len(digits) == 0 {
		return res
	}

	phone := map[byte]string{
		'2': "abc",
		'3': "def",
		'4': "ghi",
		'5': "jkl",
		'6': "mno",
		'7': "pqrs",
		'8': "tuv",
		'9': "wxyz",
	}

	path := []byte{}
	var backtrack func(index int)
	backtrack = func(index int) {
		if len(path) == len(digits) {
			res = append(res, string(path))
			return
		}
		choices := phone[digits[index]]
		for i := 0; i < len(choices); i++ {
			path = append(path, choices[i])
			backtrack(index + 1)
			path = path[:len(path)-1]
		}
	}
	backtrack(0)
	return res
}

File: daily/Backtracking/17-Letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        res = []
        if len(digits) == 0:
            return res
        phone_numbers = {
            "2": "abc",
            "3": "def",
            "4": "ghi",
            "5": "jkl",
            "6": "mno",
            "7": "pqrs",
            "8": "tuv",
            "9": "wxyz"
        }
        def backtrack(path, index):
            if len(path) == len(digits):
                letters = "".join(path)
                res.append(letters)
                return
            choices = phone_numbers[digits[index]]
            for i in range(len(choices)):
                path.append(choices[i])
                backtrack(path, index + 1)
                path.pop()
        backtrack([], 0)
        return res